src/lucas_kanade.jl

```diff
@@ -1,259 +1,240 @@
 
-
-
-function declare_lost!(status, flow, j)
-    status[j] = false
-    flow[j] = SVector{2}(0.0,0.0)
+function optflow(
+    first_img::AbstractMatrix{T}, second_img::AbstractMatrix{T},
+    points::Vector{SVector{2, Float64}}, algorithm::LucasKanade,
+) where T <: Gray
+    displacement = fill(SVector{2, Float64}(0.0, 0.0), length(points))
+    first_pyramid = LKPyramid(first_img, algorithm.pyramid_levels)
+    second_pyramid = LKPyramid(
+        second_img, algorithm.pyramid_levels; compute_gradients=false,
+    )
+    optflow!(first_pyramid, second_pyramid, points, displacement, algorithm)
 end
 
-function construct_pyramids(first_img, second_img, pyramid_levels)
-    first_pyramid = gaussian_pyramid(first_img, pyramid_levels, 2, 1.0)
-    second_pyramid = gaussian_pyramid(second_img, pyramid_levels, 2, 1.0)
-    return first_pyramid, second_pyramid
-end
-
-function initialise_arrays(N::Int)
-    flow = Array{SVector{2, Float64}, 1}(undef, N)
-    for i in eachindex(flow)
-        flow[i] = SVector{2, Float64}(0.0, 0.0)
-    end
-    status = trues(N)
-    reliability = zeros(Float64, N)
-    return flow, status, reliability
+function optflow!(
+    first_img::AbstractMatrix{T}, second_img::AbstractMatrix{T},
+    points::Vector{SVector{2, Float64}}, displacement::Vector{SVector{2, Float64}},
+    algorithm::LucasKanade,
+) where T <: Gray
+    first_pyramid = LKPyramid(first_img, algorithm.pyramid_levels)
+    second_pyramid = LKPyramid(
+        second_img, algorithm.pyramid_levels; compute_gradients=false,
+    )
+    optflow!(first_pyramid, second_pyramid, points, displacement, algorithm)
 end
 
-function compute_partial_derivatives(Ix, Iy)
-    Ixx = imfilter(Ix.*Ix, Kernel.gaussian(4))
-    Iyy = imfilter(Iy.*Iy, Kernel.gaussian(4))
-    Ixy = imfilter(Ix.*Iy, Kernel.gaussian(4))
-
-    Ixx_integral_table = integral_image(Ixx)
-    Ixy_integral_table = integral_image(Ixy)
-    Iyy_integral_table = integral_image(Iyy)
-    return Ixx_integral_table, Ixy_integral_table, Iyy_integral_table
-end
-
-
-function compute_flow_vector(corresponding_point, G, A, Ix, Iy, grid, offsets, etp)
-    # Solve a linear system of equations in order to determine the flow vector.
-    bk = prepare_linear_system(corresponding_point, A, Ix, Iy, grid, offsets, etp)
-    G_inv = pinv2x2(G) #TODO: Add the pinv2x2 method to the StaticArrays package.
-    ηk = G_inv*bk
-end
-
-function determine_pyramid_coordinates(points, i, j)
-    px = floor(Int,(points[j][1]) / 2^i)
-    py = floor(Int,(points[j][2]) / 2^i)
-    point = SVector{2}(px ,py)
-end
+function optflow!(
+    first_pyramid::LKPyramid, second_pyramid::LKPyramid,
+    points::Vector{SVector{2, Float64}}, displacement::Vector{SVector{2, Float64}},
+    algorithm::LucasKanade,
+)
+    enough_layers = (
+        length(first_pyramid.layers) > algorithm.pyramid_levels &&
+        length(second_pyramid.layers) > algorithm.pyramid_levels
+    )
+    !enough_layers && throw("Not enough layers in pyramids.")
```

Comment:
A style suggestion and feel free to ignore.

```suggestion
    has_enough_layers = length(first_pyramid.layers) > algorithm.pyramid_levels &&
        length(second_pyramid.layers) > algorithm.pyramid_levels
    has_enough_layers || throw("Not enough layers in pyramids.")
```